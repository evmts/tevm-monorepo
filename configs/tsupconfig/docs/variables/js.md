**@tevm/tsupconfig** ∙ [README](../README.md) ∙ [API](../API.md)

***

[API](../API.md) > js

# Variable: js

> **`const`** **js**: `Options`

## Source

[js.js:3](https://github.com/evmts/tevm-monorepo/blob/main/configs/tsupconfig/src/js.js#L3)

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
