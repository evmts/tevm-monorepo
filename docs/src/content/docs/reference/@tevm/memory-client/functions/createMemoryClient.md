---
editUrl: false
next: false
prev: false
title: "createMemoryClient"
---

> **createMemoryClient**(`options`?): `Promise`\<[`MemoryClient`](/reference/tevm/memory-client/type-aliases/memoryclient/)\>

A local EVM instance running in JavaScript. Similar to Anvil in your browser

## Parameters

â–ª **options?**: [`BaseClientOptions`](/reference/tevm/base-client/type-aliases/baseclientoptions/)= `{}`

## Returns

## Example

```ts
import { createMemoryClient } from "tevm"
import { MyERC721 } from './MyERC721.sol'

const tevm = createMemoryClient({
	fork: {
	  url: "https://mainnet.optimism.io",
	},
})

const address = '0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045',

await tevm.runContractCall(
  MyERC721.write.mint({
    caller: address,
  }),
)

const balance = await tevm.runContractCall(
 MyERC721.read.balanceOf({
 caller: address,
 }),
 )
 console.log(balance) // 1n
 ```

## Source

[createMemoryClient.js:35](https://github.com/evmts/tevm-monorepo/blob/main/packages/memory-client/src/createMemoryClient.js#L35)

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
