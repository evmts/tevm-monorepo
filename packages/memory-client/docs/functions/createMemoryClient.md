**@tevm/memory-client** ∙ [README](../README.md) ∙ [API](../API.md)

***

[API](../API.md) > createMemoryClient

# Function: createMemoryClient()

> **createMemoryClient**(`options`?): `Promise`\<[`MemoryClient`](../type-aliases/MemoryClient.md)\>

A local EVM instance running in JavaScript. Similar to Anvil in your browser

## Parameters

▪ **options?**: [`CreateEVMOptions`](../type-aliases/CreateEVMOptions.md)= `{}`

## Returns

## Example

```ts
import { createMemoryClient } from "tevm"
import { createPublicClient, http } from "viem"
import { MyERC721 } from './MyERC721.sol'

const tevm = createMemoryClient({
	fork: {
	  url: "https://mainnet.optimism.io",
	},
})

const address = '0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045',

await tevm.runContractCall(
  MyERC721.write.mint({
    caller: address,
  }),
)

const balance = await tevm.runContractCall(
 MyERC721.read.balanceOf({
 caller: address,
 }),
 )
 console.log(balance) // 1n
 ```

## Source

[packages/memory-client/src/createMemoryClient.js:59](https://github.com/evmts/tevm-monorepo/blob/main/packages/memory-client/src/createMemoryClient.js#L59)

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
