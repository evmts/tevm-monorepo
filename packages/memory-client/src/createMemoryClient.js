import { createBaseClient } from '@tevm/base-client'
import {
	eip1993EventEmitter,
	ethActions,
	requestEip1193,
	tevmActions,
	tevmSend,
} from '@tevm/decorators'

/**
 * A local EVM instance running in JavaScript. Similar to Anvil in your browser
 * @param {import('@tevm/base-client').BaseClientOptions} [options]
 * @returns {import('./MemoryClient.js').MemoryClient}
 * @example
 * ```ts
 * import { createMemoryClient } from "tevm"
 * import { MyERC721 } from './MyERC721.sol'
 *
 * const tevm = createMemoryClient({
 * 	fork: {
 * 	  url: "https://mainnet.optimism.io",
 * 	},
 * })
 *
 * const address = '0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045',

 * await tevm.runContractCall(
 *   MyERC721.write.mint({
 *     caller: address,
 *   }),
 * )
 *
 * const balance = await tevm.runContractCall(
 *  MyERC721.read.balanceOf({
 *  caller: address,
 *  }),
 *  )
 *  console.log(balance) // 1n
 *  ```
 */
export const createMemoryClient = (options) => {
	return createBaseClient(options)
		.extend(tevmSend())
		.extend(eip1993EventEmitter())
		.extend(requestEip1193())
		.extend(ethActions())
		.extend(tevmActions())
}
