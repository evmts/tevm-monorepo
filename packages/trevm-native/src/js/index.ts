import { createLogger } from '@tevm/logger';
import { InternalError } from '@tevm/errors';
import type { StateManager } from '@tevm/state';
import type { Common } from '@tevm/common';
import type { Chain } from '@tevm/blockchain';

// Import native module
import { init_trevm, TrevmWrapper } from '../../dist/index.node';

export interface TrevmEvmOptions {
  common: Common;
  stateManager: StateManager;
  blockchain: Chain;
  customPrecompiles?: any[];
  profiler?: boolean;
  allowUnlimitedContractSize?: boolean;
  loggingLevel?: string;
}

export interface TrevmRunCallOptions {
  caller?: string;
  to?: string;
  data?: string;
  value?: bigint;
  gasLimit?: bigint;
  skipBalance?: boolean;
}

export interface TrevmRunCallResult {
  returnValue: string;
  executionGasUsed: string;
  error?: string;
}

export interface TrevmEvm {
  runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult>;
}

/**
 * Creates a trevm-based EVM instance
 * @param options EVM options
 * @returns A trevm EVM instance
 */
export const createTrevm = async ({
  common,
  stateManager,
  blockchain,
  customPrecompiles,
  profiler,
  allowUnlimitedContractSize,
  loggingLevel,
}: TrevmEvmOptions): Promise<TrevmEvm> => {
  const logger = createLogger({
    name: '@tevm/trevm',
    level: loggingLevel ?? 'warn',
  });
  
  logger.debug({
    allowUnlimitedContractSize,
    profiler,
    customPrecompiles: customPrecompiles?.map(c => c.address.toString()),
  });

  try {
    // Initialize the native trevm module with callbacks to the StateManager
    const nativeTrevm = await init_trevm({
      chainId: common.chainId(),
      allowUnlimitedContractSize: allowUnlimitedContractSize ?? false,
      stateCallbacks: {
        getAccount: async (address: string) => {
          try {
            const account = await stateManager.getAccount(address);
            if (!account) return null;
            
            return {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? `0x${Buffer.from(account.codeHash).toString('hex')}` : undefined,
              code: account.code ? `0x${Buffer.from(account.code).toString('hex')}` : undefined,
            };
          } catch (error) {
            logger.error(`Error in getAccount for ${address}:`, error);
            return null;
          }
        },
        putAccount: async (address: string, account: any) => {
          try {
            await stateManager.putAccount(address, {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? Buffer.from(account.codeHash.slice(2), 'hex') : undefined,
              code: account.code ? Buffer.from(account.code.slice(2), 'hex') : undefined,
            });
            return true;
          } catch (error) {
            logger.error(`Error in putAccount for ${address}:`, error);
            return false;
          }
        },
        getStorage: async (address: string, key: string) => {
          try {
            const value = await stateManager.getContractStorage(address, key);
            return value ? `0x${Buffer.from(value).toString('hex')}` : '0x0';
          } catch (error) {
            logger.error(`Error in getStorage for ${address} key ${key}:`, error);
            return '0x0';
          }
        },
        putStorage: async (address: string, key: string, value: string) => {
          try {
            await stateManager.putContractStorage(
              address,
              key,
              Buffer.from(value.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putStorage for ${address} key ${key}:`, error);
            return false;
          }
        },
        getCode: async (address: string) => {
          try {
            const code = await stateManager.getContractCode(address);
            return code ? `0x${Buffer.from(code).toString('hex')}` : '0x';
          } catch (error) {
            logger.error(`Error in getCode for ${address}:`, error);
            return '0x';
          }
        },
        putCode: async (address: string, code: string) => {
          try {
            await stateManager.putContractCode(
              address,
              Buffer.from(code.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putCode for ${address}:`, error);
            return false;
          }
        },
        getBlockHash: async (blockNumber: number) => {
          try {
            const block = await blockchain.getBlock(BigInt(blockNumber));
            return block ? `0x${Buffer.from(block.header.hash()).toString('hex')}` : null;
          } catch (error) {
            logger.error(`Error in getBlockHash for block ${blockNumber}:`, error);
            return null;
          }
        }
      }
    });

    // Return a JavaScript wrapper around the native trevm that matches the Evm interface
    return {
      async runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult> {
        try {
          return await nativeTrevm.run_call({
            caller: opts.caller,
            to: opts.to,
            data: opts.data,
            value: opts.value,
            gas_limit: opts.gasLimit,
            skip_balance: opts.skipBalance,
          });
        } catch (error) {
          logger.error('Error in runCall:', error);
          throw new InternalError(`Trevm execution error: ${error}`);
        }
      }
    };
  } catch (error) {
    logger.error('Error initializing trevm:', error);
    throw new InternalError(`Failed to initialize trevm: ${error}`);
  }
};