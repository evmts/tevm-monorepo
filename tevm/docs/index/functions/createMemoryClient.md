**tevm** ∙ [README](../../README.md) ∙ [API](../../API.md)

***

[API](../../API.md) > [index](../README.md) > createMemoryClient

# Function: createMemoryClient()

> **createMemoryClient**(`options`?): `Promise`\<`MemoryClient`\>

## Parameters

▪ **options?**: [`CreateEVMOptions`](../type-aliases/CreateEVMOptions.md)

## Source

packages/memory-client/types/createMemoryClient.d.ts:1

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
