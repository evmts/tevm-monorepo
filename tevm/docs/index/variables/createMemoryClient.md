[**tevm**](../../README.md)

***

[tevm](../../modules.md) / [index](../README.md) / createMemoryClient

# Variable: createMemoryClient

> `const` **createMemoryClient**: [`CreateMemoryClientFn`](../type-aliases/CreateMemoryClientFn.md)

Defined in: packages/memory-client/types/createMemoryClient.d.ts:297

Creates a [MemoryClient](../type-aliases/MemoryClient.md) - a fully-featured Ethereum development and testing environment.

The MemoryClient is an all-in-one solution that combines:
- A complete in-memory Ethereum Virtual Machine implementation
- Full support for viem's wallet, test, and public actions
- TEVM-specific actions for advanced state manipulation and tracing
- Automatic handling of JSON-RPC requests through viem's client interface

This provides an integrated environment for local Ethereum development with capabilities like:
- Executing and debugging smart contracts without deploying to a testnet
- Forking from existing networks with state caching for realistic testing
- Direct manipulation of blockchain state for complex test scenarios
- Complete control over mining and transaction processing
- Compatibility with standard Ethereum tooling and libraries

## Example

```typescript
import { createMemoryClient, http } from "tevm";
import { optimism } from "tevm/common";

// Create a memory client that forks from Optimism mainnet
const client = createMemoryClient({
  fork: {
    transport: http("https://mainnet.optimism.io")({}),
  },
  common: optimism,
  mining: { auto: true }, // Automatically mine blocks after transactions
});

// Wait for fork initialization to complete
await client.tevmReady();

// Use standard viem actions
const blockNumber = await client.getBlockNumber();
console.log(`Connected to Optimism block ${blockNumber}`);

// Use TEVM-specific actions
await client.tevmSetAccount({
  address: "0x123...",
  balance: 10000000000000000000n // 10 ETH
});
```

## See

 - [Client Guide](https://tevm.sh/learn/clients/)
 - [Actions Guide](https://tevm.sh/learn/actions/)
 - [Reference Docs](https://tevm.sh/reference/tevm/memory-client/functions/creatememoryclient/)
 - [Viem client docs](https://viem.sh/docs/clients/introduction)

## Key Configuration Options

```typescript
createMemoryClient({
  // Chain configuration (use tevm/common chains, not viem/chains)
  common: optimism,

  // Forking from an existing network
  fork: {
    transport: http("https://mainnet.optimism.io")({}),
    blockTag: 'latest', // or specific block number/hash
  },

  // Mining configuration
  mining: {
    auto: true,      // Auto-mine after each transaction
    interval: 5000,  // Mine blocks every 5 seconds (in ms)
  },

  // State persistence
  persister: createSyncPersister({
    storage: localStorage, // or any synchronous storage
  }),

  // Chain configuration
  hardfork: 'cancun', // Default and recommended

  // EVM execution logging
  logging: {
    logLevel: 'debug',
  },
})
```

## Actions API

MemoryClient combines multiple action types in a single interface:

### 1. TEVM-specific actions
```typescript
// Account and state manipulation
await client.tevmSetAccount({ address: "0x123...", balance: 100n });
await client.tevmGetAccount({ address: "0x123..." });

// Contract interactions
await client.tevmContract({
  abi: [...],
  functionName: "transfer",
  args: ["0x456...", 1000n]
});

// Contract deployment
const result = await client.tevmDeploy({
  abi: [...],
  bytecode: "0x...",
  args: ["Constructor", "Args"]
});

// Mining control
await client.tevmMine({ blockCount: 5, interval: 10 });

// State persistence
const state = await client.tevmDumpState();
await client.tevmLoadState(state);
```

### 2. Viem public actions
```typescript
const balance = await client.getBalance({ address: "0x123..." });
const blockNumber = await client.getBlockNumber();
const code = await client.getCode({ address: "0x123..." });
const logs = await client.getLogs({ address: "0x123..." });
```

### 3. Viem test actions
```typescript
await client.setBalance({ address: "0x123...", value: 100n });
await client.impersonateAccount({ address: "0x123..." });
await client.mine({ blocks: 1 });
await client.setStorageAt({ address, index, value });
```

### 4. Viem wallet actions
```typescript
const hash = await client.sendTransaction({
  from: "0x123...",
  to: "0x456...",
  value: 1000n
});
```

## Forking Networks

The MemoryClient can fork from any EVM-compatible network, creating a local copy that
lazily loads state from the remote network as needed:

```typescript
import { createMemoryClient, http } from "tevm";
import { optimism } from "tevm/common";

const forkedClient = createMemoryClient({
  // Fork specification
  fork: {
    transport: http("https://mainnet.optimism.io")({}),
    blockTag: '0xa6a63cd70fbbe396321ca6fe79e1b6735760c03538208b50d7e3a5dac5226435',
  },
  // Always specify chain configuration for optimal performance
  common: optimism,
});
```

The `common` object extends the viem chain interface with EVM-specific information.
Always use `tevm/common` chains rather than `viem/chains` when working with TEVM.

## Mining Modes

TEVM supports three mining modes:

```typescript
// 1. Manual mining (default)
const client = createMemoryClient({ mining: { auto: false } });
await client.sendTransaction(...);  // Transaction is pending
await client.tevmMine();            // Now transaction is processed

// 2. Auto-mining (mine after every transaction)
const autoClient = createMemoryClient({ mining: { auto: true } });
await autoClient.sendTransaction(...);  // Automatically mined

// 3. Interval mining (mine periodically)
const intervalClient = createMemoryClient({
  mining: { interval: 5000 }  // Mine every 5 seconds
});
```

## Server Mode

TEVM can be exposed as an HTTP JSON-RPC server with `@tevm/server`:

```typescript
import { createServer } from "tevm/server";
import { createMemoryClient } from "tevm";

const memoryClient = createMemoryClient();
const server = createServer({
  request: memoryClient.request,
});

server.listen(8545, () => console.log("TEVM running on port 8545"));
```

This allows any Ethereum tool or library to connect to your TEVM instance:

```typescript
import { createPublicClient, http } from "viem";
import { mainnet } from "viem/chains";

const publicClient = createPublicClient({
  chain: mainnet,
  transport: http("http://localhost:8545"),
});

console.log(await publicClient.getChainId());
```

## State Persistence

TEVM state can be persisted between sessions with the `persister` option:

```typescript
import { createMemoryClient } from "tevm";
import { createSyncPersister } from "tevm/sync-storage-persister";

// Browser example with localStorage
const browserClient = createMemoryClient({
  persister: createSyncPersister({
    storage: localStorage,
    key: 'my-tevm-state'
  }),
});

// Node.js example with file system
import { FileStorage } from "tevm/sync-storage-persister/node";
const nodeClient = createMemoryClient({
  persister: createSyncPersister({
    storage: new FileStorage('./tevm-state.json'),
  }),
});
```

## Direct Solidity Imports

When used with the TEVM bundler plugins, you can import Solidity files directly:

```typescript
import { createMemoryClient } from "tevm";
import { MyERC721 } from './MyERC721.sol';

const client = createMemoryClient();

// Deploy the contract
const deployed = await client.tevmDeploy(
  MyERC721.deploy("My NFT", "NFT")
);
await client.tevmMine();

// Create contract instance with the deployed address
const nft = MyERC721.withAddress(deployed.createdAddress);

// Call contract methods
await client.tevmContract({
  ...nft.write.mint('0x123...', 1),
  from: '0x123...',
});

await client.tevmMine();

const balance = await client.tevmContract(nft.read.balanceOf('0x123...'));
console.log(balance); // 1n
```

This requires setting up one of the TEVM bundler plugins (vite, webpack, esbuild, etc.).

## Network Support

TEVM officially supports:
- Ethereum mainnet and testnets
- OP Stack chains (Optimism, Base, etc.)

Other EVM chains are likely to work but not officially tested. Chain configuration
is available through `tevm/common`.

## Advanced EVM Features

TEVM includes advanced EVM features, with the following enabled by default:
- EIP-1559 Fee Market
- EIP-4895 (Beacon chain withdrawals)
- EIP-4844 (Blob transactions)
- EIP-4788 (Beacon root in EVM)

## Optimizing Bundle Size

For UI applications concerned with bundle size, use tree-shakeable actions with `createTevmNode()`
and individual actions from `tevm/actions`. See the [actions API guide](https://tevm.sh/learn/actions/)
for details.
