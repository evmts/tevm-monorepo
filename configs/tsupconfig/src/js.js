import { createTsUpOptions } from './createTsupOptions.js'

export const js = createTsUpOptions({
	target: 'js',
})
