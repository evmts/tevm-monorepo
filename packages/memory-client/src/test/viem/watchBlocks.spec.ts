import { SimpleContract } from '@tevm/test-utils'
import { beforeEach, describe, expect, it } from 'vitest'
import { createMemoryClient } from '../../createMemoryClient.js'

let mc = createMemoryClient()

beforeEach(async () => {
	mc = createMemoryClient()
	const deployResult = await mc.tevmDeploy({
		bytecode: SimpleContract.bytecode,
		abi: SimpleContract.abi,
		args: [420n],
	})
	if (!deployResult.createdAddress) {
		throw new Error('contract never deployed')
	}
	if (!deployResult.txHash) {
		throw new Error('txHash not found')
	}
	await mc.tevmMine()
})

describe('watchBlocks', () => {
	it('should work', async () => {
		const expectedBlock = Promise.withResolvers()
		const unwatch = mc.watchBlocks({
			poll: true,
			pollingInterval: 100,
			onBlock: (block) => {
				block.hash = '0xredacted'
				block.stateRoot = '0xredacted'
				block.parentHash = '0xredacted'
				block.timestamp = 69n
				expectedBlock.resolve(block)
			},
		})
		await mc.tevmMine()
		expect(await expectedBlock.promise).toMatchSnapshot()
		unwatch()
	})
})
