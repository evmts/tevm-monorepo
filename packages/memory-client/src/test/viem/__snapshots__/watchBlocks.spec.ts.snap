// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`watchBlocks > should work 1`] = `
{
  "baseFeePerGas": 7n,
  "blobGasUsed": 0n,
  "difficulty": 0n,
  "excessBlobGas": 0n,
  "extraData": "0x",
  "gasLimit": 30000000n,
  "gasUsed": 0n,
  "hash": "0xredacted",
  "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "miner": "0x0000000000000000000000000000000000000000",
  "mixHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "nonce": "0x0000000000000000",
  "number": 2n,
  "parentBeaconBlockRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "parentHash": "0xredacted",
  "receiptsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "requests": undefined,
  "requestsRoot": undefined,
  "sha3Uncles": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  "size": 1532n,
  "stateRoot": "0xredacted",
  "timestamp": 69n,
  "totalDifficulty": 0n,
  "transactions": [],
  "transactionsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "uncles": [],
  "withdrawals": [],
  "withdrawalsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
}
`;
